### PR TITLE
Create Docker Build Stage for Front End Assets

### PR DESCRIPTION
**Pull request by alexm118**
Looking to remove some of our frontend twistlock alerts by excluding dependencies from the production build.

## Additions

- cfgov-build image as a middle layer between dev and prod

## Changes

- Change where we install node/yarn and compile static assets.

## How to test this PR

1. Lets see if it ends up running in the dev cluster

## Notes and todos

-

## Checklist

- [ ] PR has an informative and human-readable title
  - PR titles are used to generate the change log in [releases](../../releases); good ones make that easier to scan.
  - Consider prefixing, e.g., "Mega Menu: fix layout bug", or "Docs: Update Docker installation instructions".
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code follows the standards laid out in the [CFPB development guidelines](https://github.com/cfpb/development)
- [ ] Future todos are captured in comments and/or tickets
- [ ] Project documentation has been updated, potentially one or more of:
  - [This repo’s docs](https://cfpb.github.io/cfgov-refresh/) (edit the files in the `/docs` folder) – for basic, close-to-the-code docs on working with this repo
  - CFGOV/platform wiki on GHE – for internal CFPB developer guidance
  - CFPB/hubcap wiki on GHE – for internal CFPB design and content guidance

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -53,6 +53,27 @@ EXPOSE 8000
...
+# See .dockerignore for details on which files are included
+COPY . .
+
+RUN yum -y install nodejs yarn && \
+    ./frontend.sh production && \
+    cfgov/manage.py collectstatic && \
+    yarn cache clean && \
+    rm -rf node_modules npm-packages-offline-cache
+
+
```
Comment (alexm118) [inline Dockerfile:76]: This creates a container that installs node and yarn and builds the static assets, it also cleans up node_modules and npm-packages which can probably be deleted as we don't copy them into our production container.

File: Dockerfile [hunk 2]
```diff
@@ -83,11 +104,13 @@ RUN yum -y install ${SCL_HTTPD_VERSION} ${SCL_PYTHON_VERSION}-mod_wsgi && \
     echo "source scl_source enable ${SCL_HTTPD_VERSION}" > /etc/profile.d/enable_scl_httpd.sh && \
     echo '[ -d /var/run/secrets ] && cd /var/run/secrets && for s in *; do export $s=$(cat $s); done && cd -' > /etc/profile.d/secrets_env.sh
 
-# See .dockerignore for details on which files are included
-COPY --chown=apache:apache . .
+# Copy the cfgov directory form the build image
+COPY --from=cfgov-build --chown=apache:apache ${CFGOV_PATH}/cfgov ${CFGOV_PATH}/cfgov
+COPY --from=cfgov-build --chown=apache:apache ${CFGOV_PATH}/docker-entrypoint.sh ${CFGOV_PATH}/docker-entrypoint.sh
+COPY --from=cfgov-build --chown=apache:apache ${CFGOV_PATH}/static.in ${CFGOV_PATH}/static.in
```
Comment (alexm118) [inline Dockerfile:110]: Copy over just the files we need from the build container

- cfgov
- docker-entrypoint.sh
- static.in

File: Dockerfile [hunk 3]
```diff
@@ -99,12 +122,9 @@ USER apache
 # Build frontend, cleanup excess file, and setup filesystem
 # - cfgov/f/ - Wagtail file uploads
 # - /tmp/eregs_cache/ - Django file-based cache
-RUN ./frontend.sh production && \
-    cfgov/manage.py collectstatic && \
-    yarn cache clean && \
-    ln -s ${SCL_HTTPD_ROOT}/etc/httpd/modules ${APACHE_SERVER_ROOT}/modules && \
+RUN ln -s ${SCL_HTTPD_ROOT}/etc/httpd/modules ${APACHE_SERVER_ROOT}/modules && \
```
Comment (alexm118) [inline Dockerfile:125]: No longer need to run the build in the production container.

File: Dockerfile [hunk 4]
```diff
@@ -53,6 +53,27 @@ EXPOSE 8000
...
+# Build Frontend Assets
+FROM cfgov-dev as cfgov-build
+
+ENV STATIC_PATH ${APP_HOME}/cfgov/static/
+ENV PYTHONPATH ${APP_HOME}/cfgov
+
+# Django Settings
+ENV DJANGO_SETTINGS_MODULE cfgov.settings.production
+ENV DJANGO_STATIC_ROOT ${STATIC_PATH}
+ENV ALLOWED_HOSTS '["*"]'
```
Comment (willbarton) [inline Dockerfile:65]: Is there any way to avoid the duplication of these variables between the `cfgov-dev`, `cfgov-build`, and `cfgov-prod`?

Comment (alexm118) [reply Dockerfile:65]: I can play with it - I think if I define them all in `cfgov-dev` they should propogate.

Comment (alexm118) [reply Dockerfile:65]: although - I think some of them need to be duplicated, like the django settings module as we don't want to overrite the dev value.

Comment (willbarton) [reply Dockerfile:65]: Yeah. We just duplicate a few of them in the Dockerfile. But if there's not an easy/obvious way to reuse them, I don't want that question to block this.

Comment (alexm118) [reply Dockerfile:65]: @willbarton upon first attempt looks like it doesn't like passing them down through the parent - sepcifically it wasn't able to find the `STATIC_ROUTE` shared variable when it is defined in the initial cfgov-dev container. For now I think limited repeatability is ok

File: Dockerfile [hunk 5]
```diff
@@ -53,6 +53,27 @@ EXPOSE 8000
...
+
+# See .dockerignore for details on which files are included
+COPY . .
+
+RUN yum -y install nodejs yarn && \
+    ./frontend.sh production && \
+    cfgov/manage.py collectstatic && \
+    yarn cache clean && \
+    rm -rf node_modules npm-packages-offline-cache
```
Comment (Scotchester) [inline Dockerfile:74]: Are these removal steps necessary if we're only copying over the things we need explicitly? Could take a non-trivial amount of time, given the size and number of files.

**Pull request merged**